contracts/metatx/ERC2771Forwarder.sol

```diff
@@ -26,6 +26,10 @@ import {Address} from "../utils/Address.sol";
  * transactions in the mempool. In these cases the recommendation is to distribute the load among
  * multiple accounts.
  *
+ * NOTE: The `receive()` function is not available with ERC2771 because it requires `msg.data` to
+ * always have the `from` address at the end of the calldata, whereas the `receive()` function only
+ * works when data is empty.
+ *
```

Comment:
receive is can still be implemented and called if the sender is not the forwarder. 

Is in interresting to note that there is indeed an issue if the forwarder relays an empty "data" field, because in that case the caller will be interpreted as a selector. Maybe the Forwarder should require that `request.data.length >= 4`

Reply:
I very much agree that calldata size dependency might be an issue, amazing catch!

I'll improve the note and solve this but let me know if you think it can be improved further

Reply:
Reopening to make this explicit and make sure we agree:

>Maybe the Forwarder should require that `request.data.length >= 4`

I initially agreed with this idea, but I discussed it with @ernestognw and he made an interesting point that it would be very much Solidity and "Solidity ABI encoding" specific, by caring about 4-byte selectors and receive functions, whereas the rest of the contract is agnostic to Solidity specifics.

I think it's a good point so I would not add this check. It can be implemented at the request-generation and/or forwarder level.

Reply:
> Something interesting could be to add a reverting receive() implementation in ERC2771Context. It's a bit radical because it would cause receives to revert also when the sender is not the forwarder. This just makes sure that someone using ERC2771Context doesn't also implement a receive function expecting it to work with forwarded requests. What do you think?

I guess that receive would be virtual so that devs that override it with some checks. It's not a bad idea, but I'm worried it will be difficult to understand for devs.

Reply:
Although radical I'd say it's more common to write positive test cases rather than negatives. For example, instead of writing a test proving a contract can't receive funds, developers will more usually write a spec for `should receive funds when ...`, which will fail.

It's definitely less likely to make a mistake imo. Though, I agree with @Amxx on overriding. However, a messy override would require mixing the forwarder with more than 2 inherited contracts with a `receive` function, and I don't think anyone would get a complex inheritance in a forwarder.